fs::copy() unix: set file mode early

haraldh opened the pull request:
A convenience method like fs::copy() should try to prevent pitfalls a
normal user doesn't think about.

In case of an empty umask, setting the file mode early prevents
temporarily world readable or even writeable files,
because the default mode is 0o666.

In case the target is a named pipe or special device node, setting the
file mode can lead to unwanted side effects, like setting permissons on
`/dev/stdout` or for root setting permissions on `/dev/null`.

copy_file_range() returns EINVAL, if the destination is a FIFO/pipe or
a device like "/dev/null", so fallback to io::copy, too.

Fixes: https://github.com/rust-lang/rust/issues/26933
Fixed: https://github.com/rust-lang/rust/issues/37885

rust-highfive commented:
r? @withoutboats

(rust_highfive has picked a reviewer for you, use r? to override)

haraldh commented:
r? @alexcrichton

bors commented:
:umbrella: The latest upstream changes (presumably #58208) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented on src/libstd/sys/unix/fs.rs:
```diff
@@ -869,18 +870,35 @@ pub fn copy(from: &Path, to: &Path) -> io::Result<u64> {
+    // prevent users from setting permissions on e.g. `/dev/stdout` or a named pipe
+    if writer_metadata.is_file() {
+        // set the correct file mode
+        cvt_r(|| unsafe { libc::fchmod(writer.as_raw_fd(), mode) })?;
```
I think this is the same thing as `writer.set_permissions`, right?

haraldh replied:
you are right

alexcrichton commented on src/libstd/sys/unix/fs.rs:
```diff
@@ -869,18 +870,35 @@ pub fn copy(from: &Path, to: &Path) -> io::Result<u64> {
     };
 
+    let mut writer = OpenOptions::new()
+        // prevent world readable/writeable file in case of empty umask
```
I find this a bit cryptic, but could this be expanded to explain what's going on here?

alexcrichton commented on src/libstd/sys/unix/fs.rs:
```diff
@@ -869,18 +870,35 @@ pub fn copy(from: &Path, to: &Path) -> io::Result<u64> {
+
+    let writer_metadata = writer.metadata()?;
+    // prevent root from setting permissions on e.g. `/dev/null`
+    // prevent users from setting permissions on e.g. `/dev/stdout` or a named pipe
```
Sort of like above, I find this a bit cryptic. Could you expand a bit as to what's going on here?

alexcrichton commented on src/libstd/sys/unix/fs.rs:
```diff
@@ -915,21 +933,20 @@ pub fn copy(from: &Path, to: &Path) -> io::Result<u64> {
                         // - Kernel version is < 4.5 (ENOSYS)
                         // - Files are mounted on different fs (EXDEV)
                         // - copy_file_range is disallowed, for example by seccomp (EPERM)
+                        // - source or destination cannot be used with copy_file_range()
```
The others definitely make sense to me but is `EINVAL` something that legitimately comes up? That seems sort of like a big hammer to disallow such a wide range of possible errors causing them to stop using the more optimized pass.

In any case could the documentation here be updated to indicate which error is associated with this case?

haraldh commented:
new refined version with better comments and logic

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/181521136) ([raw log](https://api.travis-ci.com/v3/job/181521136/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:0a8d1876:start=1551426417322061806,finish=1551426536556350387,duration=119234288581
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
[00:04:51]    Compiling panic_unwind v0.0.0 (/checkout/src/libpanic_unwind)
[00:04:57] error: unused `core::result::Result` that must be used
[00:04:57]    --> src/libstd/sys/unix/fs.rs:903:9
[00:04:57]     |
[00:04:57] 903 |         writer.set_permissions(perm);
[00:04:57]     |
[00:04:57]     = note: `-D unused-must-use` implied by `-D warnings`
[00:04:57]     = note: `-D unused-must-use` implied by `-D warnings`
[00:04:57]     = note: this `Result` may be an `Err` variant, which should be handled
[00:04:57] error: aborting due to previous error
[00:04:57] 
[00:04:57] error: Could not compile `std`.
[00:04:57] 
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit fb98ca716fca8d9342f4cf308622a8e78c072fb7 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton fb98ca716fca8d9342f4cf308622a8e78c072fb7 -->

haraldh commented:
updated with refactored fix for Linux and other unices

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 5cf58328d16e3da737152db49d072449ea20e1b0 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 5cf58328d16e3da737152db49d072449ea20e1b0 -->

bors commented:
:hourglass: Testing commit 5cf58328d16e3da737152db49d072449ea20e1b0 with merge 526d081cd8cac2557a619c62b7b6c840a8e761cd...
<!-- homu: {"type":"BuildStarted","head_sha":"5cf58328d16e3da737152db49d072449ea20e1b0","merge_sha":"526d081cd8cac2557a619c62b7b6c840a8e761cd"} -->

bors commented:
:broken_heart: Test failed - [checks-travis](https://travis-ci.com/rust-lang/rust/builds/104640907)
<!-- homu: {"type":"BuildFailed","builder_url":"https://travis-ci.com/rust-lang/rust/builds/104640907","builder_name":"checks-travis"} -->

rust-highfive commented:
The job `armhf-gnu` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/185349399) ([raw log](https://api.travis-ci.com/v3/job/185349399/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[01:43:21] ---- [run-pass] run-pass/paths-containing-nul.rs stdout ----
[01:43:21] 
[01:43:21] error: test run failed!
[01:43:21] status: exit code: 101
[01:43:21] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/remote-test-client" "run" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/paths-containing-nul/a"
[01:43:21] ------------------------------------------
[01:43:21] ------------------------------------------
[01:43:21] uploaded "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/paths-containing-nul/a", waiting for result
[01:43:21] ------------------------------------------
[01:43:21] stderr:
[01:43:21] ------------------------------------------
[01:43:21] ------------------------------------------
[01:43:21] thread 'main' panicked at 'copy2 returned a strange NotFound on a path with NUL', /checkout/src/test/run-pass/paths-containing-nul.rs:12:23
[01:43:21] 
[01:43:21] ------------------------------------------
[01:43:21] 
[01:43:21] thread '[run-pass] run-pass/paths-containing-nul.rs' panicked at 'explicit panic', src/tools/compiletest/src/runtest.rs:3325:9
---
[01:43:21] 
[01:43:21] thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:496:22
[01:43:21] 
[01:43:21] 
[01:43:21] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/arm-unknown-linux-gnueabihf/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/run-pass" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass" "--stage-id" "stage2-arm-unknown-linux-gnueabihf" "--mode" "run-pass" "--target" "arm-unknown-linux-gnueabihf" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--linker" "arm-linux-gnueabihf-gcc" "--host-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/arm-unknown-linux-gnueabihf/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--llvm-version" "8.0.0\n" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--remote-test-client" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/remote-test-client" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[01:43:21] 
[01:43:21] 
[01:43:21] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --target arm-unknown-linux-gnueabihf
[01:43:21] Build completed unsuccessfully in 1:40:12
---
travis_time:end:1a6a3c63:start=1552694271800101533,finish=1552694271811990349,duration=11888816
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:1d6c7ac8
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:00b79145
travis_time:start:00b79145
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:2df1e748
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

kennytm commented:
@bors r-

```
[01:43:21] thread 'main' panicked at 'copy2 returned a strange NotFound on a path with NUL', /checkout/src/test/run-pass/paths-containing-nul.rs:12:23
```

bors commented:
:umbrella: The latest upstream changes (presumably #59226) made this pull request unmergeable. Please resolve the merge conflicts.

haraldh commented:
Hmm, I cannot reproduce this error... Strange

alexcrichton commented:
It may be how that test is emulated perhaps? It's sort of executing in a weird environment. The test assertion may also just want to be updated because it may be hitting an error at a different time than before

haraldh commented:
that should do the trick 😄 

```
    let _ = cstr(from)?;
    let _ = cstr(to)?;
```

except the error path the compiler should remove the unneeded code in the optimization

alexcrichton commented:
I think in this case it's probably more appropriate to just update the test, exactly what error comes out isn't something that we guarantee

haraldh commented:
Allright, fixed the macos version to use `fcopyfile` and "fixed" the testsuite

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 9100593e6f26a7840ac13b34ad67f21222cc1780 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 9100593e6f26a7840ac13b34ad67f21222cc1780 -->

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/186404456) ([raw log](https://api.travis-ci.com/v3/job/186404456/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:1a938dc9:start=1553097330414502875,finish=1553097454419417380,duration=124004914505
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
$ export GCP_CACHE_BUCKET=rust-lang-ci-cache
Setting environment variables from .travis.yml
---
[01:03:32] .................................................................................................... 1800/2959
[01:03:41] .................................................................................................... 1900/2959
[01:03:53] .........i......................................................................i................... 2000/2959
[01:04:14] .................................................................................................... 2100/2959
[01:04:33] ..F................................................................................................. 2200/2959
[01:04:54] ................................................ii.................................................. 2400/2959
[01:05:07] .................................................................................................... 2500/2959
[01:05:26] .................................................................................................... 2600/2959
[01:05:43] .................................................................................................... 2700/2959
---
[01:06:09] 
[01:06:09] ------------------------------------------
[01:06:09] stderr:
[01:06:09] ------------------------------------------
[01:06:09] thread 'main' panicked at 'Failed to create empty file `a`: Os { code: 26, kind: Other, message: "Text file busy" }', src/libcore/result.rs:997:5
[01:06:09] 
[01:06:09] ------------------------------------------
[01:06:09] 
[01:06:09] thread '[run-pass] run-pass/paths-containing-nul.rs' panicked at 'explicit panic', src/tools/compiletest/src/runtest.rs:3325:9
---
[01:06:09] 
[01:06:09] thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:496:22
[01:06:09] 
[01:06:09] 
[01:06:09] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/run-pass" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "run-pass" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-6.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "6.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[01:06:09] 
[01:06:09] 
[01:06:09] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:06:09] Build completed unsuccessfully in 0:09:08
[01:06:09] Build completed unsuccessfully in 0:09:08
[01:06:09] make: *** [check] Error 1
[01:06:09] Makefile:48: recipe for target 'check' failed
bj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:02b1b03c
travis_time:start:02b1b03c
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:2742495d
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

haraldh commented:
hmpf
```
[01:06:09] thread 'main' panicked at 'Failed to create empty file `a`: Os { code: 26, kind: Other, message: "Text file busy" }', src/libcore/result.rs:997:5
```

kennytm commented:
@bors r=alexcrichton

bors commented:
:pushpin: Commit d997eac498ba064b4115fb15fe33a685ef562238 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton d997eac498ba064b4115fb15fe33a685ef562238 -->

bors commented:
:hourglass: Testing commit d997eac498ba064b4115fb15fe33a685ef562238 with merge fb56242a681cd7997fca4065570b7ae73cb94c66...
<!-- homu: {"type":"BuildStarted","head_sha":"d997eac498ba064b4115fb15fe33a685ef562238","merge_sha":"fb56242a681cd7997fca4065570b7ae73cb94c66"} -->

bors commented:
:broken_heart: Test failed - [checks-travis](https://travis-ci.com/rust-lang/rust/builds/105521759)
<!-- homu: {"type":"BuildFailed","builder_url":"https://travis-ci.com/rust-lang/rust/builds/105521759","builder_name":"checks-travis"} -->

rust-highfive commented:
Your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/187173206) ([raw log](https://api.travis-ci.com/v3/job/187173206/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

haraldh commented:
Hmpf, this filesystem test runs as WASM

haraldh commented:
added
```
// ignore-emscripten no files
```
to test

haraldh commented:
removed the ignore and added
```
// If `dummy_file` does not exist, then we might get another unrelated error
#[cfg(not(target_os = "emscripten"))] let dummy_file = std::env::current_exe().unwrap();
#[cfg(target_os = "emscripten")] let dummy_file = "a";
```

haraldh commented:
Another alternative would be to use TempDir as `src/test/run-pass-fulldeps/rename-directory.rs` does.

haraldh commented:
Or just `RUST_TEST_TMPDIR` as `src/test/run-pass-fulldeps/create-dir-all-bare.rs`

alexcrichton commented:
Hm I wonder if this test could just be ignored on emscripten? You can just add a `#[cfg(not(...))]` annotation to the whole test, and it's not too important to bend over backwards to get it running on emscripten.

haraldh commented:
> Hm I wonder if this test could just be ignored on emscripten? You can just add a `#[cfg(not(...))]` annotation to the whole test, and it's not too important to bend over backwards to get it running on emscripten.

Basically I did that with the latest version. For emscripten it's the old unchanged test (which should pass then) and for all the others it is hopefully picking up the current executable as the test file.

alexcrichton commented:
Oh sure yeah and this might pass, but it's pretty jarring to be reading a test and randomly see a conditional statement for one particular platform. It's pretty common, however, to ignore tests on platforms for various reasons, so I think it's best to just ignore the whole test

haraldh commented:
> Oh sure yeah and this might pass, but it's pretty jarring to be reading a test and randomly see a conditional statement for one particular platform. It's pretty common, however, to ignore tests on platforms for various reasons, so I think it's best to just ignore the whole test

I didn't want to take away the test for emscripten, because it seems to pass in the past.

alexcrichton commented:
Let's please ignore the test for emscripten, because we don't really need to keep tests running that just happened to pass before.

haraldh commented:
> Let's please ignore the test for emscripten, because we don't really need to keep tests running that just happened to pass before.

here we go, although that contradicts my understanding of a CI

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit cf8347ba6bc82c41de2ad9bf561af593a89cbe45 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton cf8347ba6bc82c41de2ad9bf561af593a89cbe45 -->

haraldh commented:
Halleluja 😀
